piecrust/src/store/memory.rs

```diff
@@ -186,3 +181,288 @@ impl From<Memory> for VMMemory {
+        let file = fs::OpenOptions::new()
+            .create(true)
+            .read(true)
+            .write(true)
+            .open(file)?;
+
+        let file_len = file.metadata()?.len();
+        if file_len % WASM_PAGE_SIZE as u64 != 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "File length is not a multiple of WASM page size",
+            ));
+        }
+
+        // Set the length of the file to the minimum if too small.
+        let min_len = MIN_MEM_SIZE as u64;
+        if file_len < min_len {
+            file.set_len(min_len)?;
+        }
+
+        let fd = file.as_raw_fd();
+        let file_len = cmp::max(file_len, min_len) as usize;
+
+        let ptr = mmap(
+            ptr::null_mut(),
+            MAX_MEM_SIZE,
+            PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_NORESERVE,
+            fd,
+            0,
+        )?;
+
+        let growth_ptr = unsafe { ptr.add(file_len) };
+
+        mmap(
+            growth_ptr,
+            MAX_MEM_SIZE - file_len,
+            PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE | MAP_FIXED,
+            -1,
+            0,
+        )?;
+
+        Ok(Self { ptr, len: file_len })
+    }
+
+    /// Grow the map by the specified number of bytes.
+    ///
+    /// Growing in this scheme effectively means incrementing the length.
+    pub fn grow_by(&mut self, delta_bytes: usize) -> io::Result<()> {
+        let new_len = self.len + delta_bytes;
+
+        if new_len > MAX_MEM_SIZE {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("Tried growing past the maximum number of bytes: {MAX_MEM_SIZE}"),
+            ));
+        }
+
+        self.len = new_len;
+
+        Ok(())
+    }
+
+    /// Return the memory as a byte slice.
+    pub fn as_bytes(&self) -> &[u8] {
+        let ptr = self.ptr as *const u8;
+        unsafe { slice::from_raw_parts(ptr, self.len) }
+    }
+
+    /// Return the mmap as a mutable byte slice.
+    pub fn as_bytes_mut(&mut self) -> &mut [u8] {
+        unsafe { slice::from_raw_parts_mut(self.ptr, self.len) }
+    }
+}
+
+impl Deref for MemoryMmap {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.as_bytes()
+    }
+}
+
+impl DerefMut for MemoryMmap {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.as_bytes_mut()
+    }
+}
+
+impl LinearMemory for MemoryMmap {
+    fn ty(&self) -> MemoryType {
+        MemoryType {
+            minimum: Pages(WASM_MIN_PAGES as u32),
+            maximum: Some(Pages(WASM_MAX_PAGES - 1)), // there is one guard page
+            shared: false,
+        }
+    }
+
+    fn size(&self) -> Pages {
+        let pages = self.len / WASM_PAGE_SIZE - 1; // there is one guard page
+        Pages(pages as u32)
+    }
+
+    fn style(&self) -> MemoryStyle {
+        MemoryStyle::Static {
+            bound: Pages(WASM_MAX_PAGES),
```

Comment:
Should be "-1" due to guard page?

Reply:
The WASM spec says that bound is *all mapped pages*. That's why I used the full memory size here.